### PR TITLE
[WIP] FastBoot compatibility 

### PR DESCRIPTION
**Pull request by sivakumar-kailasam**
**Pages:**
- [x] Index
- [x] Install
- [x] Logout
- [x] keywords 
- [x] individual keyword
- [ ]  Individual Crate 
- [ ] search
- [ ] Login
- [ ] User

### CODE REVIEW

File: app/controllers/application.js [hunk 1]
```diff
@@ -52,17 +57,23 @@ export default Ember.Controller.extend({
...
+        }
     },
 
     stepFlash() {
-        this.set('flashError', this.get('nextFlashError'));
-        this.set('nextFlashError', null);
+        this.setProperties({
+            'flashError': this.get('nextFlashError'),
+            'nextFlashError': null
+        });
```
Comment (Turbo87) [inline app/controllers/application.js]: how is that related to fastboot?

File: app/mixins/fastboot-utils.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    appURL: computed(function() {
+        let url = '';
+        if (this.get('fastboot.isFastBoot')) {
+            let protocol = this.get('fastboot.request.protocol');
+            let host = this.get('fastboot.request.host');
+            url = `${protocol}://${host}`;
+        }
+        return url;
+    }),
```
Comment (Turbo87) [inline app/mixins/fastboot-utils.js]: comment here might be valuable to understand why we need this

File: app/routes/logout.js [hunk 3]
```diff
@@ -1,8 +1,11 @@
 import Ember from 'ember';
+import fetch from 'fetch';
+import FastbootUtils from '../mixins/fastboot-utils';
+
+export default Ember.Route.extend(FastbootUtils, {
 
-export default Ember.Route.extend({
     activate() {
-        Ember.$.getJSON('/logout', () => {
+        fetch(`${this.get('appURL')}/logout`, () => {
```
Comment (Turbo87) [inline app/routes/logout.js]: should probably use `ajax()` for consistency for now

File: config/environment.js [hunk 4]
```diff
@@ -23,6 +23,10 @@ module.exports = function(environment) {
         APP: {
             // Here you can pass flags/options to your application instance
             // when it is created
+        },
+
+        fastboot: {
+            hostWhitelist: ['crates.io', /^localhost:\d+$/]
```
Comment (Turbo87) [inline config/environment.js]: can we add localhost only in dev mode?

File: app/controllers/application.js (hunk 1 above)
Comment (sivakumar-kailasam) [inline app/controllers/application.js]: Not related. I just grouped couple of individual set calls. Should've probably kept in a different commit.

File: app/mixins/fastboot-utils.js [hunk 5]
```diff
@@ -8,6 +8,12 @@ export default Ember.Mixin.create({
 
     isNotFastBoot: computed.not('fastboot.isFastBoot'),
 
+    /**
+     * When there's a need to raise network requests from the server
+     * during Server Side Rendering(SSR) via FastBoot,
+     * the JS Fetch API doesn't work with absolute urls.
+     * This property gives the URL prefix for those network calls.
+     * */
```
Comment (Turbo87) [inline app/mixins/fastboot-utils.js]: isn't it the other way around? it doesn't work with relative URLs, right?

Comment (sivakumar-kailasam) [inline app/mixins/fastboot-utils.js]: danngit, you're right. I don't think we need it anymore if I use ajax service.

File: app/controllers/me/index.js [hunk 6]
```diff
@@ -1,20 +1,20 @@
 import Ember from 'ember';
-import ajax from 'ic-ajax';
+
+const { inject: { service } } = Ember;
```
Comment (Turbo87) [inline app/controllers/me/index.js]: this will make it harder for the new modules codemod to transform it into an import from `@ember/service`... I'd rather just keep using the globals for now.

Comment (sivakumar-kailasam) [inline app/controllers/me/index.js]: Oh so that's why you've used full form everywhere. Got it!

Comment (Turbo87) [inline app/controllers/me/index.js]: yep, and I don't multi-level destructuring that much 😉

**Comment by sivakumar-kailasam** [comment]
Closing this PR in favor of smaller PRs to get this repo to be fastboot compatible and then land one PR with the fastboot specific changes.
